[숫자 야구 게임] 방예혁 미션 제출합니다.

YehyeokBang opened the pull request:
## 질문
- `View`의 책임을 결정하는 것이 어려웠습니다. 
예시1) `InputView`에서 입력을 받고 검증까지 해야하는지? 저희는 Number 객체를 생성할 때 검증하도록 구현했습니다.
예시2) 스트라이크와 볼 결과를 출력할 때 `OutputView`에게 스트라이크와 볼 개수를 주고 메시지를 만들게 하는 것이 괜찮은지? 아니라면 어디선가 완성된 문자열을 전달해야하는지? 등
- 컴퓨터 숫자(난수)를 만들 때 정적 메서드를 사용하는건 어떤지? 그 객체가 값을 가지고 있는 것은 아니고 난수를 만들고 반환만 한다면 정적 메서드도 좋은가요? 정적 메서드는 어떨 때 사용하는지가 궁금해요.
- 아래의 코드처럼 객체를 생성할 때 new 키워드를 사용해서 객체를 주입하는 방식은 안좋아보이는데 다른 방법이 있나요? 사실 좋은 방법인지 아닌지도 잘 모르겠습니다.
```java
public class BaseballGameController {

    private final InputView inputView;
    private final OutputView outputView;
    private final Computer computer;

    public BaseballGameController() {
        this.inputView = new InputView();
        this.outputView = new OutputView();
        this.computer = new Computer();
    }
    
    ...
}
``` 

## 다음 제출까지 고민하는 부분
- 현재는 `Controller`에 있는 숫자 비교 로직 분리
- 숫자 비교 로직 효율 고민 (볼, 스트라이크 카운팅)
- 의미 있는 이름의 상수로 대체될 수 있는 수 또는 문자열 변경
- 객체 주입 관련 고민

## 기타
- 페어와 일정이 안맞아서 3시간만에 구현하다보니 부족한 점이 많을 것 같습니다...!

hangillee committed:
Update README.md

hangillee committed:
Update README.md

YehyeokBang committed:
docs: 구현할 기능 목록 추가

- 입력, 출력, 기능으로 분류하여 과제 진행합니다.

inhooo00 committed:
docs: 구현할 입력 기준 추가

inhooo00 committed:
feat: 게임 시작 메세지 출력 기능 구현

YehyeokBang committed:
feat: 숫자 입력 기능 구현

- 숫자 입력 메시지를 출력합니다.
- 기준에 맞지 않는 입력을 받은 경우 예외가 발생합니다.
- 숫자를 가진 객체 내부에서 값을 검증합니다.

inhooo00 committed:
feat: 컴퓨터 난수 생성 기능 구현

- 숫자 관리를 문자열 기반에서 리스트 기반으로 변경합니다.

YehyeokBang committed:
feat: 게임 결과 출력 기능 구현

YehyeokBang committed:
fix: 입력받은 값이 아스키 값으로 반환되는 로직 수정

YehyeokBang committed:
feat: 야구 게임 진행 기능 구현

- 재시작 로직은 아직 미구현 상태입니다.

inhooo00 committed:
feat: 재시작 기능 구현

- 1을 입력하면 재시작합니다.

hangillee commented on docs/README.md:
```diff
@@ -0,0 +1,25 @@
+    - [X]  나머지 경우
+- [x]  게임 종료
+    - [x]  재시작
+    - [x]  종료
```
기능 구현 목록 작성 👍

hangillee commented on src/main/java/baseball/controller/BaseballGameController.java:
```diff
@@ -0,0 +1,73 @@
+        while (true) {
+            Number number = inputView.getInputNumber();
+
+            Result result = compare(number.getNumber(), computer.getNumber());
+
+            outputView.printResultMessage(result.strike(), result.ball());
```
DTO를 활용하셨네요! DTO가 필요한 이유가 무엇인가요? 그냥 `model` 계층의 객체들에게서 바로 값을 가져오면 안 되나요? 🤔

YehyeokBang replied:
DTO를 사용하여 필요한 데이터(`strike`와 `ball`)를 한 번에 전달하여 코드 간결하게(불필요한 메서드 호출 감소) 만들기 위해 사용했습니다!

hangillee replied:
DTO는 여러 데이터를 한 번에 넘기는데 유용한 점도 있지만, 가장 중요한 장점은 핵심 비즈니스 로직이나 민감한 정보를 담고 있는 Domain을 다른 계층으로부터 완전히 분리시킬 수 있다는 점입니다! 👍

hangillee commented on src/main/java/baseball/controller/BaseballGameController.java:
```diff
@@ -0,0 +1,73 @@
+
+            outputView.printResultMessage(result.strike(), result.ball());
+
+            if (result.strike() == 3) {
+                if (inputView.isRestart()) {
+                    computer.generateRandomNumber();
+                    continue;
+                }
+                break;
```
`early return`을 활용해봅시다.
차라리 `break`가 먼저 호출 되었다면 `if`문의 `indent`(들여쓰기)가 한 단계여도 되지 않았을까요?

YehyeokBang replied:
넵 수정해보겠습니다!

hangillee commented on src/main/java/baseball/controller/BaseballGameController.java:
```diff
@@ -0,0 +1,73 @@
+    public BaseballGameController() {
+        this.inputView = new InputView();
+        this.outputView = new OutputView();
+        this.computer = new Computer();
```
저는 문제 없어 보입니다. 다형성을 위해 의존성 주입을 활용한다면 모르겠으나, 지금 당장은 해당 객체들의 다형성을 활용해야할 마땅한 이유가 없어 보입니다. 오히려, `InputView`와 `OutputView`를 `static` 메소드만 가지도록 만들어 Util 클래스로 바꾸면 `BaseballGameController`가 가질 상태의 개수도 줄어들 것 같습니다.

hangillee commented on src/main/java/baseball/controller/BaseballGameController.java:
```diff
@@ -0,0 +1,73 @@
+        set.addAll(computerNumber);
+
+        return 6 - set.size();
+    }
```
이 친구들은 Controller가 가져야 할 책임일까요? 비즈니스 로직에 가까워 보이는데 오히려 도메인이 가져야 할 친구들 아닐까요? Controller의 역할과 책임의 범위는 무엇일까요? 예혁이 생각하는 Controller의 역할과 책임에 대해서 알려주세요!

또, 예혁이 고민한 대로 이 친구를 분리해주세요! (힌트를 드리자면 Computer가 가지는 것도 나쁘지 않아 보여요.)

hangillee commented on src/main/java/baseball/dto/Result.java:
```diff
@@ -0,0 +1,7 @@
+        int strike,
+        int ball
+) {
+}
```
`record` 활용 👍

hangillee commented on src/main/java/baseball/model/Computer.java:
```diff
@@ -0,0 +1,31 @@
+
+    public void generateRandomNumber() {
+        List<Integer> computer = new ArrayList<>();
+        while (computer.size() < 3) {
```
매직 넘버를 상수화 해볼까요?

hangillee commented on src/main/java/baseball/model/Computer.java:
```diff
@@ -0,0 +1,31 @@
+            int randomNumber = Randoms.pickNumberInRange(1, 9);
+            if (!computer.contains(randomNumber)) {
+                computer.add(randomNumber);
+            }
```
메소드로 분리할 수 있을 것 같아요. 그렇게 하면 들여쓰기 단계도 줄일 수 있고, 메소드 이름만 보고, 자세한 조건과 동작을 모르고도 어떤 값들이 리스트에 추가될 지 유추할 수 있도록 리팩토링 해주세요!

hangillee commented on src/main/java/baseball/model/Computer.java:
```diff
@@ -0,0 +1,31 @@
+        generateRandomNumber();
+    }
+
+    public void generateRandomNumber() {
```
이 친구를 정적 메소드로 만들었을 때 얻는 이점이 무엇이라고 생각하시나요? Computer 객체가 유틸 클래스가 아닌 이상, 정적 메소드로 만들어야 할 이유가 있을까요?

YehyeokBang replied:
`Computer` 객체 생성 없이 `Computer.generateRandomNumber()` 처럼 간편하게 호출이 가능하기 때문에 코드가 간결해지고 정적 메소드는 객체마다 생성되지 않고 클래스에 속하기 때문에 메모리 사용량을 줄일 수 있을 것 같습니다. 특히, `Computer` 객체가 단순히 `generateRandomNumber()` 메소드만 제공하는 경우에는 메모리 효율성 향상 효과가 더 클 것 같습니다.

hangillee replied:
반대로 숫자 야구 게임의 규모가 커져서 여러 숫자 야구 게임 방이 생기고 참가자가 많아지면 프로그램이 종료될 때까지 `Computer.generateRandomNumber()`가 메모리에서 사라지지 않고 자리를 축내고 있을 가능성이 높지 않을까요?

`static`을 사용한다고 해서 메모리 효율성 효과가 더 클 것 같진 않아요. 심지어, 여러 사람이 동시에 숫자 야구를 플레이하게 될 경우에는 동시성 문제도 발생할 수 있을 것 같아요. GC의 대상도 되지 않을테니 메모리 누수의 위험도 있습니다!

YehyeokBang replied:
여러 방면을 생각하게 되네요 감사합니다!

hangillee commented on src/main/java/baseball/model/Number.java:
```diff
@@ -0,0 +1,68 @@
+    }
+
+    private void validation(String inputNumber) {
+        if (!isNaturalNumber(inputNumber)) {
+            throw new IllegalArgumentException("1부터 9까지의 자연수를 입력해주세요.");
+        }
+        if (!isThreeDigitNumber(inputNumber)) {
+            throw new IllegalArgumentException("3자리수를 입력해주세요.");
+        }
+        if (!isDuplicate(inputNumber)) {
+            throw new IllegalArgumentException("중복되지 않은 수를 입력해주세요.");
+        }
```
각 `if`문들을 메소드로 분리하여 메소드 이름을 통해 의미를 부여해주세요. 그렇게되면 `validation` 메소드를 대강 훑어 보더라도 어떤 내용들을 검증하는지 화인할 수 있을거에요!

hangillee commented on src/main/java/baseball/model/Number.java:
```diff
@@ -0,0 +1,68 @@
+    }
+
+    private boolean isNaturalNumber(String inputNumber) {
+        for (char c : inputNumber.toCharArray()) {
+            if (c < '1' || c > '9') {
+                return false;
+            }
```
정규식을 활용한다면...? `Stream`도 활용해볼 수 있을 것 같네요! 😁

YehyeokBang replied:
정규식을 활용하면 개발자가 더 읽기 좋은 코드가 될 것 같습니다.

hangillee commented on src/main/java/baseball/model/Number.java:
```diff
@@ -0,0 +1,68 @@
+    }
+
+    private boolean isDuplicate(String inputNumber) {
+        Set<Character> set = new HashSet<>();
```
컬렉션의 특징을 활용한 중복 제거 👍

hangillee commented on src/main/java/baseball/view/InputView.java:
```diff
@@ -0,0 +1,24 @@
+        System.out.print("숫자를 입력해주세요 : ");
+        String input = Console.readLine();
+
+        return new Number(input);
```
`View`가 `Model` 패키지의 객체를 참조하게 되었네요. MVC 패턴으로 나눈 이유가 사라진 것 같아요. 리팩토링 해봅시다!

hangillee commented on src/main/java/baseball/view/OutputView.java:
```diff
@@ -0,0 +1,27 @@
+        }
+        if (strike > 0 && ball == 0) {
+            sb.append(strike).append("스트라이크");
+            if (strike == 3) {
+                sb.append("\n3개의 숫자를 모두 맞히셨습니다! 게임 종료");
+            }
```
이 친구를 가장 상위로 올리고 `return` 해버리면 아래의 조건들은 참조할 필요도 없지 않을까요?

hangillee commented on src/main/java/baseball/view/OutputView.java:
```diff
@@ -0,0 +1,27 @@
+        System.out.println(sb);
+    }
+
+}
```
EOL에 대해서 학습해봅시다!

hangillee commented on src/test/java/baseball/ApplicationTest.java:
```diff
@@ -21,13 +21,37 @@ class ApplicationTest extends NsTest {
 
     @Test
-    void 예외_테스트() {
+    void 세자리가_아닌_숫자_입력_예외_테스트_1() {
+        assertSimpleTest(() ->
+                assertThatThrownBy(() -> runException("12"))
+                        .isInstanceOf(IllegalArgumentException.class)
+        );
+    }
+
+    @Test
+    void 세자리가_아닌_숫자_입력_예외_테스트_2() {
         assertSimpleTest(() ->
                 assertThatThrownBy(() -> runException("1234"))
                         .isInstanceOf(IllegalArgumentException.class)
         );
     }
 
+    @Test
+    void 중복된_숫자_입력_예외_테스트() {
+        assertSimpleTest(() ->
+                assertThatThrownBy(() -> runException("111"))
+                        .isInstanceOf(IllegalArgumentException.class)
+        );
+    }
+
+    @Test
+    void 자연수가_아닌_입력_예외_테스트() {
+        assertSimpleTest(() ->
+                assertThatThrownBy(() -> runException("abc"))
+                        .isInstanceOf(IllegalArgumentException.class)
+        );
+    }
+
```
테스트 작성 👍

hangillee reviewed (changes requested):
안녕하세요, 예혁! 👋

시간도 모자라고 바쁜 와중에 미션 하느라 고생하셨습니다. 그래도 이번 미션을 한 번 경험해봤던 덕분인지 훌륭하게 요구사항을 충족하신 것 같습니다!

남겨주신 질문에 대한 답변입니다!
> View의 책임을 결정하는 것이 어려웠습니다.

예혁이 생각하는 `View`의 책임은 무엇인가요? 저는 `View`는 사용자와 프로그램의 소통을 위한 인터페이스라고 생각합니다. 사용자의 입력값을 `View`에서 검증하는 것도 자연스럽고, 해당 값을 통해 객체를 생성하는 부분에서 검증하는 것도 자연스럽다고 생각합니다.

다만, `Model`은 웬만해서는 `View`로부터 넘어온 값과 분리되어야 합니다. 즉, `View`의 수정에 `Model`이 영향을 받으면 안 된다는 의미입니다. 또한, 출력 메시지도 값만 넘겨받고 `View`가 만들게 하는게 더 낫습니다. 앞서 설명한대로 `View`의 변경이 다른 계층까지 전이되어서는 안 되니까요!

> 컴퓨터 숫자(난수)를 만들 때 정적 메서드를 사용하는건 어떤지? 그 객체가 값을 가지고 있는 것은 아니고 난수를 만들고 반환만 한다면 정적 메서드도 좋은가요? 정적 메서드는 어떨 때 사용하는지가 궁금해요.

정적 메소드는 클래스 로드와 동시에 메모리에 올라가기 때문에 클래스를 인스턴스화 하지 않아도 사용 가능합니다. 즉, 인스턴스화 하지 않아도 모든 클래스에서 접근 가능하기 때문에 오용될 가능성이 높아집니다. 그래서 정적 메소드는 다른 객체와 협력하기 보단 스스로 기능을 제공하는 유틸 클래스에서 사용됩니다. `Math` 클래스가 대표적이죠.

정적 메소드의 장단은 명확하기 때문에 예혁의 판단하에 필요하다면 사용해도 좋습니다. 다만, 단점이 무엇이고 그로 인해 발생하는 문제가 무엇인지만 잘 알고 있다는 전제하에 사용해야 합니다.

> 아래의 코드처럼 객체를 생성할 때 new 키워드를 사용해서 객체를 주입하는 방식은 안좋아보이는데 다른 방법이 있나요? 사실 좋은 방법인지 아닌지도 잘 모르겠습니다.

객체를 생성하는 방식은 개발자가 결정하기 나름입니다. 사실 지금 Controller 코드는 다형성을 사용했을 때 마땅히 이점을 얻을만한 객체도 안 보이고 외부에서 의존성 주입을 할만한 코드도 보이지 않습니다. (오히려 `main` 메소드에서 `View`와 `Model`을 아는 게 더 이상해 보입니다.) 한가지 개선점을 찾자면, 굳이 생성자에서 `new`를 통해 인스턴스화 할 이유가 없다는 것입니다. 그냥 필드에 바로 `new` 해주면 안되나요? 어차피 결과는 같습니다!

테스트 코드도 추가로 작성하고 여러모로 미션을 잘 수행해주셔서 감사합니다! 코멘트 남겨드린 부분만 리팩토링 해봅시다! 힌트를 드리자면, 메소드 이름만 보고도 어떤 역할을 하는지 알 수 있어야 합니다. 흔히, **축약 금지**라고 불리는 규칙입니다. 역할과 책임도 한 번 재고해봅시다!

YehyeokBang committed:
docs: 1차 리뷰 후 개선 방향 추가

inhooo00 committed:
feat: 1차 리뷰 후 개선 방향 추가

: 하드코딩 수정, Integer 통일

YehyeokBang committed:
refactor: 사용자가 입력한 숫자 관련 모델 수정

- 전달받은 String 객체가 null인지 확인하는 절차를 추가합니다.
- matches() 메서드와 정규식을 사용하여 자연수인지 확인합니다.
- Stream API를 사용하여 전달받은 값을 리스트로 변환합니다.
- 확인하는 메서드들을 정적 메서드로 변경합니다.

inhooo00 committed:
docs: 오탈자 수정

inhooo00 committed:
refactor: 코드 리팩토링

- 숫자 비교 로직을 모델로 이동합니다.
- 최대한 한 개의 책임을 가지도록 메서드를 분리합니다.

YehyeokBang committed:
refactor: 추가 리팩토링

- 매직 넘버, 문자열을 상수화합니다.
- 일부 조건문들을 early return으로 변경합니다.

hangillee commented on src/main/java/baseball/constant/GameRule.java:
```diff
@@ -0,0 +1,9 @@
+    public static final int MIN_NUMBER = 1;
+    public static final int MAX_NUMBER = 9;
+    public static final int NUMBERS_SIZE = 3;
+
```
공백 라인을 둔 특별한 이유가 있나요?

YehyeokBang replied:
위 3개는 사용자가 게임 내에 사용하는 숫자(범위, 개수)와 관련되었고, 아래는 게임 재시작과 관련된 값이라서 고민하다가 띄워두었습니다.

hangillee commented on src/main/java/baseball/constant/GameRule.java:
```diff
@@ -0,0 +1,9 @@
+package baseball.constant;
+
+public class GameRule {
```
`private` 생성자를 추가해주세요! 인스턴스화를 막아야 할 것 같습니다!

hangillee commented on src/main/java/baseball/constant/ViewMessage.java:
```diff
@@ -0,0 +1,15 @@
+package baseball.constant;
+
+public class ViewMessage {
```
마찬가지로 `private` 생성자를 선언해주세요!

YehyeokBang replied:
넵

hangillee commented on src/main/java/baseball/controller/BaseballGameController.java:
```diff
@@ -0,0 +1,44 @@
+
+        inputView.printGameStart();
+
+        while (true) {
+            game(computer);
+
+            if (!inputView.isRestart()) {
+                break;
+            }
+
+            computer.generateUniqueRandomNumbers();
+        }
+    }
```
코드를 조금만 바꾼다면 `indent`가 1로 줄어들 것 같습니다!

힌트) `inputView.isRestart()`의 반환값을 변수에 받아온다면...?

YehyeokBang replied:
고민해보겠습니다!

hangillee commented on src/main/java/baseball/controller/BaseballGameController.java:
```diff
@@ -0,0 +1,44 @@
+    }
+
+    private void game(Computer computer) {
+        while (true) {
+            String input = inputView.getInputNumber();
+            Numbers numbers = new Numbers(input);
+
+            Result result = computer.calculateScore(numbers.getNumbers());
+
+            outputView.printResultMessage(result);
+
+            if (result.isGameEnd()) {
+                break;
+            }
+        }
```
이 친구도 마찬가지로 `indent`를 1로 줄일 수 있을 것 같아요. 리팩토링 해봅시다.

hangillee commented on src/main/java/baseball/dto/Result.java:
```diff
@@ -0,0 +1,10 @@
+        int strike,
+        int ball
+) {
+    public boolean isGameEnd() {
+        return this.strike == 3;
+    }
```
DTO에 비즈니스 로직에 두는 것은 위험할 수도 있습니다! 지금은 간단하지만 더 많은 로직이 생길 경우 문제가 생길 여지가 있습니다. DTO가 다양한 도메인에서 사용될 때, 다른 개발자가 로직을 오용하기 쉬울 수 있고, 의존하는 도메인이 많아질 수록 해당 로직을 변경하는데 부담이 커집니다. 해당 로직을 다른 객체에 적절하게 위임해봅시다!

YehyeokBang replied:
넵!

hangillee commented on src/main/java/baseball/constant/ViewMessage.java:
```diff
@@ -0,0 +1,15 @@
+package baseball.constant;
+
+public class ViewMessage {
```
`Enum`을 활용해볼 수도 있을 것 같아요!

YehyeokBang replied:
처음에는 `Enum`을 사용해서 만들었는데, `Enum`도 객체이기 때문에 선언된 값(문자열, 정수 등)을 사용하려면 아래처럼 메서드 하나를 더 써야했습니다. 제가 구현한 코드에서는 `Enum`을 사용해서 얻는 이점보다 깔끔하게 보이는 것이 더 좋다고 생각해서 이렇게 구현했습니다. 현재 코드에서 `Enum` 사용했을 때의 이점이 무엇이 있는지 궁금합니다..! 또한 이렇게 간단한 구현에서 비교 후 선택한 제 행동이 괜찮은 접근인지 궁금합니다!
```java
// Enum 사용
String msg = ViewMessage.GAME_START.getMessage();

// 현재
String msg = ViewMessage.GAME_START;
```

hangillee replied:
`Enum`을 사용하면 `Enum` 자체에 로직을 두어 스스로 어떤 로직을 처리하게 둘 수도 있습니다. 또, `values()` 메소드를 활용해 `Enum`의 구체적인 값을 모르고도 로직을 처리하는 방법도 있습니다. 자세한 내용은 제 체스 게임 repository의 `Enum`들을 확인해보시면 됩니다!

hangillee replied:
물론, 예혁의 이런 접근도 전 괜찮다고 생각합니다. 당장에 `ViewMessage` 객체는 자체적으로 수행할 로직이 없어 보이니 이대로 둬도 무방할 것 같습니다.

확고한 자신만의 기준 👍

hangillee commented on src/main/java/baseball/model/Computer.java:
```diff
@@ -0,0 +1,64 @@
+    public void generateUniqueRandomNumbers() {
+        Set<Integer> uniqueNumbers = new HashSet<>();
+
+        while (uniqueNumbers.size() < 3) {
```
매직 넘버 `3`대신 `NUMBERS_SIZE`를 사용하는 건 어떨까요?

hangillee commented on src/main/java/baseball/model/Computer.java:
```diff
@@ -0,0 +1,64 @@
+
+import static baseball.constant.GameRule.MAX_NUMBER;
+import static baseball.constant.GameRule.MIN_NUMBER;
+import static baseball.constant.GameRule.NUMBERS_SIZE;
```
`Enum`이나 상수에 대한 `static import`에 대해 코드의 가독성을 저해시키는 요인 중 하나라는 의견이 있습니다. `static import`를 하지 않는 것을 추천드립니다.

hangillee commented on src/main/java/baseball/model/Computer.java:
```diff
@@ -0,0 +1,64 @@
+        return countDuplicateNumber(this.numbers, userNumbers);
+    }
+
+    // Set 자료구조를 사용해 중복되는 수의 개수를 반환하는 메서드
```
주석 없어도 충분히 이해가게 잘 작성해주셨어요! 지워주시면 좋을 것 같습니다.

hangillee commented on src/main/java/baseball/model/Computer.java:
```diff
@@ -0,0 +1,64 @@
+    private int countDuplicateNumber(List<Integer> list1, List<Integer> list2) {
+        int totalSize = list1.size() + list2.size();
+
+        Set<Integer> set = new HashSet<>();
+        set.addAll(list1);
+        set.addAll(list2);
```
컬렉션을 적극 활용하는 코드 👍

hangillee commented on src/main/java/baseball/model/Numbers.java:
```diff
@@ -0,0 +1,66 @@
+            throw new IllegalArgumentException("중복되지 않는 수를 입력해주세요.");
+        }
+
+        // GC가 처리하도록 명시적으로 null로 설정
+        set = null;
```
GC가 처리하도록 신경쓰지 않아도 될 것 같아요. 이 정도는 메모리에 크게 영향은 가지 않습니다. 이전 피드백 때문에 해두신 건 이해하지만 굳이 필요한 코드는 아닌 것 같습니다. 오히려 주석이 없으면 의도를 이해하기 힘든 코드라 컨벤션으로 정한게 아닌 이상 다른 개발자가 코드를 이해하는데 방해가 될 수 있습니다.

hangillee commented on src/main/java/baseball/model/Numbers.java:
```diff
@@ -0,0 +1,66 @@
+        return inputNumbers.chars()
+                .filter(Character::isDigit)
+                .mapToObj(c -> (Character.digit(c, 10)))
+                .collect(Collectors.toList());
```
그냥 `.toList()`로 사용할 수 있습니다!

YehyeokBang replied:
변경될 이유가 없으니 불변 리스트로 반환하는 것이 더 좋겠네요 감사합니다!

hangillee reviewed (changes requested):
리팩토링 수고하셨습니다!
조금만 더 수정하면 될 것 같아서 코멘트 남깁니다.

이번 리뷰는 페어와 같이 진행할 필요는 없습니다.

YehyeokBang reviewed (commented):
감사합니다. 더 고민해보겠습니다.

YehyeokBang committed:
refactor: 추가 리팩토링

- View의 메서드를 정적 메서드로 변경합니다.
- indent를 최소화하려고 했습니다.
- DTO가 로직을 가지지 않도록 수정합니다.
- 숫자 중복 확인 로직을 for문에서 Stream을 사용하도록 수정합니다.

hangillee commented on src/main/java/baseball/constant/GameRule.java:
`GameRule`은 `model` 패키지에, `ViewMessage`는 `view` 패키지로 이동하는게 좋아보여요. 패키지 의존성도 중요합니다.

hangillee reviewed (approved):
안녕하세요, 예혁! 👋

피드백 잘 반영해주셨네요. 몇 가지만 더 신경쓰면 좋을 것 같습니다. POSIX New Line에 대해서 학습해보고 다음 미션부터는 신경써주세요! merge하도록 하겠습니다!

자동차 경주 미션도 파이팅!
